### PR TITLE
Restart TPLs: new checkpoints.period syntax

### PR DESCRIPTION
**Pull request by wmles**
Write a script for computing next step from the new-style period syntax.
Necessary for restart.tpl, which has to extract the parameters for the next restart.

### CODE REVIEW

**Comment by wmles** [comment]
@PrometheusPi  bitte gucken!!!!!!! ;)

File: etc/picongpu/hypnos-hzdr/k80_restart.tpl [hunk 1]
```diff
@@ -99,30 +99,60 @@ do
...
-else
-    echo "no checkpoint found" | tee -a output
-    endTime=$restartPeriod
-    stepSetup=$(echo " -s " $endTime "--checkpoint.period" $restartPeriod )
-fi
+# ******************************************* #
+# need some magick, if the restart period is in new notation with the ':' and ','
+
+which python3 >&2
+python3 --version >&2
```
Comment (psychocoderHPC) [inline etc/picongpu/hypnos-hzdr/k80_restart.tpl]: Sry but we should not  pull python into the templates because than you need also the correct python environment.

Comment (wmles) [reply etc/picongpu/hypnos-hzdr/k80_restart.tpl]: well, but it works :)
3 arguments:
 - if you like, we can add a module load
 - it's only k80, so none except us will have trouble if something would not work (and we are all pros and will understand ;))
 - if it doesn't work, you see it on submitting, at once - no hidden errors, no hours of lost computing time

It would probably be cleaner to put the code in a script, that accepts some parameters and returns some, shall I? and where shall the script be located?

**Comment by ax3l** [comment]
@wmles please provide a PR description and remove the white spaces that make the CI crash:
```bash
sed -i 's/[[:blank:]]\+$//' ./etc/picongpu/hypnos-hzdr/k80_restart.tpl
```

**Comment by ax3l** [comment]
@wmles is `etc/picongpu/hypnos-hzdr/k20_restart.tpl` also affected? Please update this file as well!

File: etc/picongpu/hypnos-hzdr/k80_restart.tpl [hunk 2]
```diff
@@ -99,33 +99,62 @@ do
...
+    steps += range(int(os.environ['finalStep'])+1)[slice(*parts)]
+
+steps.append(int(os.environ['finalStep']))
+print('steps-Liste: %s' % steps, file=sys.stderr)
+# return list and next bigger timestep as endpoint for next run
+if not os.environ['file']:
+    next_step = min(steps)
+else:
+    current_step = int(os.path.splitext(os.path.basename('$file'))[0].strip('checkpoint_'))
+    print('aktueller Step: %s ' % current_step, file=sys.stderr)
```
Comment (ax3l) [inline etc/picongpu/hypnos-hzdr/k80_restart.tpl]: please do not write German in code or comments.

**Comment by ax3l** [comment]
@wmles ping :)

**Comment by ax3l** [comment]
@wmles @PrometheusPi ping :)

File: etc/picongpu/hypnos-hzdr/k80_restart.tpl [hunk 3]
```diff
@@ -99,33 +99,62 @@ do
...
+    next_step = min([step for step in steps if step>current_step])
+    print('naechster Step: %s ' % next_step, file=sys.stderr)
```
Comment (n01r) [inline etc/picongpu/hypnos-hzdr/k80_restart.tpl]: more German here

File: etc/picongpu/hypnos-hzdr/k80_restart.tpl [hunk 4]
```diff
@@ -99,33 +99,62 @@ do
...
+        parts = [None, None, parts[0]]
+    elif len(parts)==2:
+        parts = [parts[0], None if parts[1] is None else parts[1]+1, None]
+    elif len(parts)==3:
+        parts = [parts[0], None if parts[1] is None else parts[1]+1, parts[2]]
+
+    steps += range(int(os.environ['finalStep'])+1)[slice(*parts)]
+
+steps.append(int(os.environ['finalStep']))
+print('steps-Liste: %s' % steps, file=sys.stderr)
```
Comment (n01r) [inline etc/picongpu/hypnos-hzdr/k80_restart.tpl]: please replace the German in here as well

File: etc/picongpu/hypnos-hzdr/k80_restart.tpl [hunk 5]
```diff
@@ -99,33 +99,62 @@ do
...
+# return list and next bigger timestep as endpoint for next run
+if not os.environ['file']:
+    next_step = min(steps)
+else:
+    current_step = int(os.path.splitext(os.path.basename('$file'))[0].strip('checkpoint_'))
```
Comment (n01r) [inline etc/picongpu/hypnos-hzdr/k80_restart.tpl]: just a remark: this only works if `--checkpoint.restart.file` hasn't been changed

Comment (PrometheusPi) [reply etc/picongpu/hypnos-hzdr/k80_restart.tpl]: That is true for the current version as well. Thanks for pointing this out.

File: src/tools/bin/nextstep_from_period.sh [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+# PIConGPU is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with PIConGPU.
+# If not, see <http://www.gnu.org/licenses/>.
+#
+
```
Comment (ax3l) [inline src/tools/bin/nextstep_from_period.sh]: Please add [`set -euf -o pipefail`](https://github.com/azet/community_bash_style_guide#style-conventions) for modern bash scripts.
Then pls verify it still works and modernize if needed.

File: src/tools/bin/nextstep_from_period.sh [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+for s in $sections
+do
+    newSteps=($(expandPeriods $s))
+    steps+=(${newSteps[@]})
+done
+
+steps=($(printf "%s\n" "${steps[@]}" | sort -u -g))
+
+for step in "${steps[@]}"; do
+    if [ "$step" -gt "$currentStep" ]; then
```
Comment (ax3l) [inline src/tools/bin/nextstep_from_period.sh]: pls don't put these numbers in `""`

File: src/tools/bin/nextstep_from_period.sh [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        echo "$step"
```
Comment (ax3l) [inline src/tools/bin/nextstep_from_period.sh]: no `""` needed

File: src/tools/bin/nextstep_from_period.sh [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+# Script returns next timestep specified by the period syntax
+#
+# Takes three lines with values from stdin:
+# - a string representing period-syntax, like `3000,12000:15000:1000,1337:1337`
+# - the current timestep (convertible to int)
+# - and the final timestep (convertible to int)
+#
+# and prints the next timestep to stdout
+
+read period
```
Comment (ax3l) [inline src/tools/bin/nextstep_from_period.sh]: I think simple args syntax with a `-h|--help` implemented on wrong number of arguments would be more user friendly

File: src/tools/bin/nextstep_from_period.sh [hunk 10]
```diff
@@ -19,22 +19,52 @@
...
+
+case $# in
+  0)
+    printf "\n *** Not enough arguments! *** \n"
+    helpText
+    exit 1
+  ;;
+  1)
+    helpText
+    if [ "$1" = "-h" ] || [ "$1" = "-h" ]; then
```
Comment (PrometheusPi) [inline src/tools/bin/nextstep_from_period.sh]: question:
what is the difference between both or-cases?
Should it be instead as follows?
```bash
if [ "$1" = "-h" ] || [ "$1" = "--help" ];
```

File: etc/picongpu/hypnos-hzdr/k80_restart.tpl [hunk 11]
```diff
@@ -142,7 +143,7 @@ fi
 
 mpiexec --prefix $MPIHOME -x LIBRARY_PATH -npernode !TBG_gpusPerNode -n !TBG_tasks /usr/bin/env bash -c "killall -9 picongpu 2>/dev/null || true"
 
-if [ $endTime -lt $finalStep ]
+if [ $nextStep -lt $finalStep ]
```
Comment (PrometheusPi) [inline etc/picongpu/hypnos-hzdr/k80_restart.tpl]: question:
why the renaming?

File: src/tools/bin/nextstep_from_period.sh [hunk 12]
```diff
@@ -0,0 +1,108 @@
+#!/usr/bin/env bash
+#
+# Copyright 2017-2018 Axel Huebl
```
Comment (PrometheusPi) [inline src/tools/bin/nextstep_from_period.sh]: @wmles Please add your name to the author list, or was this code entirely written by @ax3l ?

File: src/tools/bin/nextstep_from_period.sh [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+    currentStep="$3"
+  ;;
+  *)
+    printf "\n *** Too many arguments! *** \n"
+    helpText
+    exit 1
+  ;;
+esac
+
+splitSections()
```
Comment (PrometheusPi) [inline src/tools/bin/nextstep_from_period.sh]: please add comments on purpose, output, and needed arguments

File: src/tools/bin/nextstep_from_period.sh [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+{
+    IFS=','
+    for s in $1
+    do
+        echo $s
+    done
+}
+expandPeriods()
```
Comment (PrometheusPi) [inline src/tools/bin/nextstep_from_period.sh]: please add comments on purpose, output, and needed arguments

File: src/tools/bin/nextstep_from_period.sh [hunk 15]
```diff
@@ -19,22 +19,52 @@
...
-# - the current timestep (convertible to int)
-# - and the final timestep (convertible to int)
-#
-# and prints the next timestep to stdout
+set -euf -o pipefail
+
+helpText()
+{
+    echo "
+    Usage: nextstep_from period period lastStep [currentStep]
```
Comment (psychocoderHPC) [inline src/tools/bin/nextstep_from_period.sh]: `period` is two times on the line

File: src/tools/bin/nextstep_from_period.sh [hunk 16]
```diff
@@ -19,22 +19,52 @@
...
+}
+
+case $# in
+  0)
+    printf "\n *** Not enough arguments! *** \n"
+    helpText
+    exit 1
+  ;;
+  1)
+    helpText
```
Comment (psychocoderHPC) [inline src/tools/bin/nextstep_from_period.sh]: This help must be moved into the if block

Comment (wmles) [reply src/tools/bin/nextstep_from_period.sh]: no, the help text shall be printed always, the difference between "-h" and "42" ist only the exit code

File: etc/picongpu/hypnos-hzdr/k80_restart.tpl [hunk 17]
```diff
@@ -99,31 +99,32 @@ do
         file=""
     fi
 done
+export file
 
 #this sed call extracts the final simulation step from the cfg (assuming a standard cfg)
-finalStep=`echo !TBG_programParams | sed 's/.*-s[[:blank:]]\+\([0-9]\+[^\s]\).*/\1/'`
+export finalStep=`echo !TBG_programParams | sed 's/.*-s[[:blank:]]\+\([0-9]\+[^\s]\).*/\1/'`
```
Comment (ax3l) [inline etc/picongpu/hypnos-hzdr/k80_restart.tpl]: is the export of these vars and vars below really needed?

File: src/tools/bin/nextstep_from_period.sh [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+# along with PIConGPU.
+# If not, see <http://www.gnu.org/licenses/>.
+#
+
+set -euf -o pipefail
+
+helpText()
+{
+    echo "
+    Usage: nextstep_from_period.sh period lastStep [currentStep]
```
Comment (ax3l) [inline src/tools/bin/nextstep_from_period.sh]: You can use `$0` instead of `nextstep_from_period.sh`

File: src/tools/bin/nextstep_from_period.sh [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+
+    Script returns next timestep specified by the period syntax
+    Takes three positional arguments:
+     - a string representing period-syntax, like 3000,12000:15000:1000,1337:1337
+     - the current timestep (convertible to int)
```
Comment (ax3l) [inline src/tools/bin/nextstep_from_period.sh]: order is different than above here

File: etc/picongpu/hypnos-hzdr/k80_restart.tpl (hunk 17 above)
Comment (wmles) [reply etc/picongpu/hypnos-hzdr/k80_restart.tpl]: probably not; it may have appeared there while debugging (but didn't help) and stayed there, sorry

File: src/tools/bin/nextstep_from_period.sh (hunk 19 above)
Comment (wmles) [reply src/tools/bin/nextstep_from_period.sh]: ups, yeah, it's wrong; I reordered it to make the last argument optional

**Comment by ax3l** [comment]
@wmles can you please also update `etc/picongpu/hypnos-hzdr/k20_restart.tpl`?

**Comment by ax3l** [comment]
@psychocoderHPC @PrometheusPi looks good to me now as well :)

File: etc/picongpu/hypnos-hzdr/k80_restart.tpl [hunk 20]
```diff
@@ -99,31 +99,32 @@ do
         file=""
     fi
 done
+file
```
Comment (PrometheusPi) [inline etc/picongpu/hypnos-hzdr/k80_restart.tpl]: Why is this line added to the script? 
Is this a typo?

File: src/tools/bin/nextstep_from_period.sh [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+{
+    # split a string on occurences of ','
+    IFS=','
+    for s in $1; do
+        echo $s
+    done
+}
+expandPeriods()
+{
+    # split a string on occurences of ':' and always return a 3-tuple
```
Comment (PrometheusPi) [inline src/tools/bin/nextstep_from_period.sh]: It does not return a 3-tuple - it internally splits the given string into a 3-typle (with default values if needed) and returns a sequence of checkpoint time steps. 
Please specify the comment.

**Comment by ax3l** [comment]
@PrometheusPi @psychocoderHPC if you review this as well I can merge it later when it's run through the test again (just a whitespace change was left...)

File: etc/picongpu/hypnos-hzdr/k80_restart.tpl (hunk 11 above)
Comment (ax3l) [reply etc/picongpu/hypnos-hzdr/k80_restart.tpl]: looks proper to me now. it's the next step that is compared here

File: etc/picongpu/hypnos-hzdr/k80_restart.tpl (hunk 20 above)
Comment (ax3l) [reply etc/picongpu/hypnos-hzdr/k80_restart.tpl]: yes, probably accidentally pasted. not present in the other file

File: etc/picongpu/hypnos-hzdr/k80_restart.tpl (hunk 11 above)
Comment (PrometheusPi) [reply etc/picongpu/hypnos-hzdr/k80_restart.tpl]: *offline* discussion with @wmles:
both namings have their pro and cons (`Step`sound more like the next PIC iteration while `$endTime` might be interpreted as final step of all simulations.) Both are fine with me.

File: src/tools/bin/nextstep_from_period.sh (hunk 21 above)
Comment (ax3l) [reply src/tools/bin/nextstep_from_period.sh]: will fix

**Comment by ax3l** [comment]
@PrometheusPi pushed fixes

**Comment by PrometheusPi** [comment]
@ax3l Thanks for the quick fix.
